### DISCUSSION TITLE
Future plans, OC2R and general alignment

### DISCUSSION THREAD

**Discussion by Leonardo5681** [fnuecke/oc2#295] [General]
I know it's not really a report, but I look forward to the moment when this mod will be stable, yet this moment isn't coming, what happened to this mod?

**Comment by CoolCat467** [comment]
Basically on hiatus at the moment it seems. [There was an issue with the underlying emulator](https://github.com/fnuecke/sedna/issues/12), but apparently that might have been fixed now, so kind of in limbo.

**Comment by samuelh2005** [comment]
[oc2r](https://github.com/North-Western-Development/oc2r) is a continuation development of this mod. Currently it supports 1.20.1 Forge, but they have plans to port to Fabric by using Architectury. 

Fnuecke did attempt to update however they never made that public.

**Comment by TheOverpassArsonist** [comment]
> [oc2r](https://github.com/North-Western-Development/oc2r) is a continuation development of this mod. Currently it supports 1.20.1 Forge, but they have plans to port to Fabric by using Architectury.
> 
> Fnuecke did attempt to update however they never made that public.

damn so 19.2 is just being skipped entirely then? Well that sucks. I get why people do it but this version stampede makes it really annoying to try to build any full modpack. It feels like 1.12 or maybe even 1.14 or 1.16 were the last times there was a real staple version. (Granted the forge/fabric split is also a large contributor here)

(to be extra clear here, no I don't blame any individual developer for this, I just find it annoying how there hasn't been a real 'staple version' in years.)

**Comment by samuelh2005** [comment]
@TheOverpassArsonist 

> damn so 19.2 is just being skipped entirely then?

Actually no, OC2R supports 1.19.2 but it is not being updated anymore. So 1.19.2 version does exist but no new features are being added.

However the JAR files are not published to any mod platform.

**Comment by fnuecke** [comment]
Life happened, as it does, repeatedly. I've had a bit of spurt now, but who knows how long that'll last. Also wasn't aware of OC2R, oh well. FWIW, the merge requests to sedna should have fixed some of the fundamental issues; things should be more stable now generally, but time will tell.

**Comment by Un1q32** [comment]
> [@TheOverpassArsonist](https://github.com/TheOverpassArsonist)
> 
> > damn so 19.2 is just being skipped entirely then?
> 
> Actually no, OC2R supports 1.19.2 but it is not being updated anymore. So 1.19.2 version does exist but no new features are being added.
> 
> However the JAR files are not published to any mod platform.

the 1.19.2 version died very soon after being created, it was only around briefly before the 1.20.1 version was created.

**Comment by Un1q32** [comment]
Will you be merging in changes from oc2r?  There's a lot of work that was done there that it would be sad to lose.  Particularly overlayfs and the terminal improvements.  There's also the internet card but that was always half baked and a nightmare to setup.

**Comment by fnuecke** [comment]
If you're up for that, I'd definitely be interested, sure! What would be your preferred mode? Pull requests? Pointing me at what you'd like to see merged so I can get a picture first?

For the internet access topic in particular though: I'll put that off for a bit longer, I'll want to get a proper understanding on the matter. And if it needs custom (possibly platform specific) config outside MC on the host system, I'd rather that be an addon mod (necessary APIs should be there I think).

For context: right now my focus is fixing the reported bugs here and going through the open pull requests. Then I'll have to think about which things from og oc I might want to add (transposer block e.g.), and after that there's some design ideas I want to experiment with in the RPC/IO device layer.

(Disclaimer: started a modded session with friends recently, so one of the topics after that might well be integration. And how long that runs will probably also influence motivation ;))

**Reply by bastetfurry** [reply]
Regarding the internet access, most of us would be fine with some SLiRP-like implementation like what 86Box does. Works on every system, needs no setup by the user and one can wget from the internet.

**Reply by fnuecke** [reply]
That's basically what it is now yeah! Had to boil down the PR a bit to wrap my head around it, in particular I kicked out the API and extensibility, it's purely internal, purely ipv4 now, but I think that's sufficient for the use case.

**Comment by fnuecke** [comment]
Ah, and continuing the reply from #236 so that we have it in one place for now:

> If you hadn't seen already, there was a continuation fork https://github.com/North-Western-Development/oc2r. I was one of the major developers on that, maintaining the linux distro https://github.com/North-Western-Development/minux. Changes include overlayfs support, which drastically increases available hard drive space by having a shared read only drive that holds the actual os stuff, while the main drive only saves changes, a much better terminal, basically fully conforming to the xterm-256color spec, many more programs work that didn't before, and the distro is actually up to date, and an update to 1.20.1. There was also a 1.21.1 version that was basically completed but there was some temporary small issue with a dependency and by the time it got fixed nobody cared enough to do the last bit to get it across the finish line. Unfortunately it eventually died when everyone involved lost interest, the only things still being done is me occasionally updating the linux distro.

Yeah, I had not, thanks for the links!

- The overlayfs sounds interesting, I'd assume that'd skip the initial copy step and avoid each rootfs taking the full 8mb blob? I'd been thinking about hand-rolling something like that [in Sedna](https://github.com/fnuecke/sedna/blob/main/src/main/java/li/cil/sedna/device/block/SparseBlockDevice.java) back then (at least I think that's what that was for), but don't think I ever got around to trying to implement it in oc2.
- Fancy terminal also sounds neat and should be fairly isolated I'd assume. Mildly worried that might conflict a bit since I've also been fixing that area up a bit.
- Re minux, what's that sitting at, image/rootfs-size-wise? I've finally managed to cram things into sizes that are "real", which I kinda like. I.e. none of the storages (flash, floppy, hdd) use "magical extra storage" right now; partly as a proof of concept that custom OSes should work, as long as they fit on the storages, partially as a matter of principle. Ideally I'd like to keep that. If more programs would crop up as wanted/useful, I was planning on adding back loot disks for those. Also the kernel finally boots on 8MB RAM. Barely, but it does.

**Comment by Un1q32** [comment]
Minux is *much* to big to fit on a drive for real, you would need overlays.

Overlayfs is a system built into Linux that we just leveraged, you don't need to add anything too complex.  It just attaches the rootfs as a read only drive always connected to every computer, and the users drive stores changes, so the whole drive appears to be entirely writable.  It also means that updates to the OS will automatically apply to old systems without having to create a new drive, which might be good or bad depending on your priorities.

I wouldn't know how to make PRs containing the changes or even where to find the relevant code, I know exactly zero Java.  My work on oc2r was entirely about maintaining the Linux distro.

**Reply by fnuecke** [reply]
Hmm, will have to think on that then, but I'll have a look when I can regarding the overlayfs at the very least.

**Comment by Un1q32** [comment]
I would highly recommend trying to merge the terminal code, I haven't tried the new version of the mod but the old version had literally the worst terminal I've ever used that wasn't like, ultra simple no color stuff.  Even relatively simple stuff like htop didn't work properly.  The new terminal handled everything I threw at it.

**Reply by fnuecke** [reply]
Ahh, got it, will have a look then when I'm done with the issue/pr stuff!

Certainly would prio the terminal though, yeah. Fixed a lot of stuff after messing around with vttest a bit, so I'm painfully aware how broken it was.

**Comment by samuelh2005** [comment]
Once all that is done, I would like to see a bridge for CC: Tweaked / Computer Craft. I think it would look like this:
- the OC computer is placed next to the CC computer
- then JSON data can be passed back and forth, perhaps fires and event on the CC side

This would avoid having to make an API for every CC peripheral and the pain of trying to implement a network stack for CC: Tweaked and allow users to make their own integration/control scripts.

I do know Java, so I could make a PR, I will have to figure out how this mod's hardware driver interfaces work.

**Comment by SquidDev** [comment]
This probably getting very off topic now, but one of the things we've discussed with OC:Rebooted is it might make more sense to have a common API for computer mod integration, and getting mods to adopt that, rather than adding extra integrations and/or trying to bridge CC peripherals.

I don't think one need something very comprehensive. Most mods just need `@LuaFunction`/`@Callback` and `IDynamicPeripheral`/`RPCDevice`/`ManagedPeripheral`.

**Reply by fnuecke** [reply]
Converted the issue into a conversation to make it a little easier to keep this organized.

Sounds great! Since the annotations are usually just reflection based niceness on top of a more generic underpinning, something like that should be pretty straightforward to integrate I think. A dedicated repo to track that standard with an embeddable set of API would be my first idea on how this could look like, pragmatically?

Main thing I'd like to put up for consideration is that if any "advanced" functionality **is** desired (I remember stuff like returned objects that have a lifetime of their own from back then), that could maybe be an explicit extension, so that the core stays simple enough to implement for a broad range of "computers". Anything extra could maybe be an optional thing, but not something to rely on to be available, basically.

**Reply by SquidDev** [reply]
> A dedicated repo to track that standard with an embeddable set of API would be my first idea on how this could look like, pragmatically?

Yep! I'm basically thinking a separate org/repo with shared control across OC2/OC:R/CC:T maintainers. Then publish a library which can be jar-in-jared by mods. To be clear, I've not got beyond the reaching-out-to-people stage — OC:R and OC2 are only newly on 1.21.1, so it's only recently been worth doing.

Definitely would want to keep the API surface relatively simple to implement. I think the MVP is basically whatever would be needed to support Create and Mekanism (I think they're the biggest content mods with CC:T integration right now).

I think there's two things Create uses right now OC2 doesn't/won't support.
 - `RPCDevice` has no way to send signals/events/interrupts to attached computers. I think this is something we could just document as explicitly optional, and suggest mods provide a way to poll for this data too.

 - Create has [one Lua object](https://github.com/Creators-of-Create/Create/blob/0924e93639ad5f61cfc39a221d909e16f2893df1/src/main/java/com/simibubi/create/compat/computercraft/implementation/luaObjects/PackageLuaObject.java) (and yes, it gets the lifetimes wrong).

   I think nested objects are useful, but they're an absolute pain when it comes to lifetimes and persistence, so definitely be in favour of making it optional, or just leaving out support entirely.

**Reply by fnuecke** [reply]
Got it, and yeah, makes sense to me! Would definitely be on board adopting that.

For events, that's on my list of things to look into. It's *technically* possible now, there's just no clean path to it from the API yet, something I have to think about a bit. I've mostly added it as proof of concept for now, only used by robots to signal they've completed an action and for device bus change notifications to allow caching it. But I think that's a reasonable thing to add as a baseline, with a note that ideally the info sent via events *should* also be available by polling.

**Comment by brisingraerowing** [comment]
One thing I'd note is that the older buildroot may be a security issue with how old some of the programs in that version are, especially if the internet gets involved. I think OC2:R using the OverlayFS and newer buildroot (Minux) is good for avoiding such issues.

Also, a way to have multiple variants of an OS that could be used in different contexts would be cool. For example, a Client variant with things like the Lynx browser, and a server variant with additional server software (PHP and the like). Making an in-game 'internet' with online shops and other websites would be really cool.
